### PR TITLE
Create next.js example for pullstate

### PR DESCRIPTION
**Pull request by xeoneux**

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| buildDuration | 8.6s | 8.6s | ⚠️ +21ms |
| nodeModulesSize | 82.5 MB | 82.5 MB | ✓ |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.081 | 2.058 | -0.02 |
| / avg req/sec | 1201.6 | 1215.06 | +13.46 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.077 | 1.098 | ⚠️ +0.02 |
| /error-in-render avg req/sec | 2321.3 | 2277.45 | ⚠️ -43.85 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| 677f882d2ed8..34b9.js gzip | 12.8 kB | 12.8 kB | ✓ |
| framework.HASH.js gzip | 39 kB | 39 kB | ✓ |
| main-3e8cb31..cddc.js gzip | 6.56 kB | 6.56 kB | ✓ |
| webpack-e067..f178.js gzip | 751 B | 751 B | ✓ |
| Overall change | 59 kB | 59 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| polyfills-4b..e242.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| _app-7231d4b..5856.js gzip | 1.28 kB | 1.28 kB | ✓ |
| _error-01375..90b6.js gzip | 3.69 kB | 3.69 kB | ✓ |
| hooks-d4591d..e7c2.js gzip | 887 B | 887 B | ✓ |
| index-17468f..5d83.js gzip | 227 B | 227 B | ✓ |
| link-db223d9..dbd7.js gzip | 1.61 kB | 1.61 kB | ✓ |
| routerDirect..fd29.js gzip | 303 B | 303 B | ✓ |
| withRouter-b..76ae.js gzip | 302 B | 302 B | ✓ |
| Overall change | 8.31 kB | 8.31 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| _buildManifest.js gzip | 322 B | 322 B | ✓ |
| Overall change | 322 B | 322 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| index.html gzip | 614 B | 614 B | ✓ |
| link.html gzip | 620 B | 620 B | ✓ |
| withRouter.html gzip | 608 B | 608 B | ✓ |
| Overall change | 1.84 kB | 1.84 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| buildDuration | 10.1s | 10.2s | ⚠️ +110ms |
| nodeModulesSize | 82.5 MB | 82.5 MB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| 677f882d2ed8..34b9.js gzip | 12.8 kB | 12.8 kB | ✓ |
| framework.HASH.js gzip | 39 kB | 39 kB | ✓ |
| main-3e8cb31..cddc.js gzip | 6.56 kB | 6.56 kB | ✓ |
| webpack-e067..f178.js gzip | 751 B | 751 B | ✓ |
| Overall change | 59 kB | 59 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| polyfills-4b..e242.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| _app-7231d4b..5856.js gzip | 1.28 kB | 1.28 kB | ✓ |
| _error-01375..90b6.js gzip | 3.69 kB | 3.69 kB | ✓ |
| hooks-d4591d..e7c2.js gzip | 887 B | 887 B | ✓ |
| index-17468f..5d83.js gzip | 227 B | 227 B | ✓ |
| link-db223d9..dbd7.js gzip | 1.61 kB | 1.61 kB | ✓ |
| routerDirect..fd29.js gzip | 303 B | 303 B | ✓ |
| withRouter-b..76ae.js gzip | 302 B | 302 B | ✓ |
| Overall change | 8.31 kB | 8.31 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| _buildManifest.js gzip | 322 B | 322 B | ✓ |
| Overall change | 322 B | 322 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| _error.js | 1 MB | 1 MB | ✓ |
| 404.html | 2.67 kB | 2.67 kB | ✓ |
| hooks.html | 1.92 kB | 1.92 kB | ✓ |
| index.js | 1 MB | 1 MB | ✓ |
| link.js | 1.06 MB | 1.06 MB | ✓ |
| routerDirect.js | 1.05 MB | 1.05 MB | ✓ |
| withRouter.js | 1.05 MB | 1.05 MB | ✓ |
| Overall change | 5.16 MB | 5.16 MB | ✓ |

</details>

</details>
Commit: 5ad15002cb990164e4a0c2872deb644a0fd8d323

File: examples/with-pullstate/pages/_app.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+import { PullstateCore } from '../stores'
+
+export default function App({ Component, pageProps }) {
+  const instance =
+    typeof window === 'undefined'
+      ? PullstateCore.instantiate({ ssr: true })
+      : PullstateCore.instantiate({
+          ssr: false,
+          hydrateSnapshot: pageProps.hydrateSnapshot,
+        })
```
Comment (lfades) [inline examples/with-pullstate/pages/_app.js:12]: Isn't this going to create a new store every time `App` re-renders?

**Review by lfades — CHANGES_REQUESTED** [review]
@xeoneux Hi there 👋  and thank you for the PR!

Let's make sure to show the store working in different rendering methods, like in the [`with-redux`](https://github.com/vercel/next.js/tree/canary/examples/with-redux) example, where the store is used for both SSR and static pages, and navigations work without any issues.

**Comment by xeoneux** [comment]
Hi @lfades thanks for the review, I have updated the example with a demo for ssg and ssr pages

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| buildDuration | 8.3s | 8.5s | ⚠️ +241ms |
| nodeModulesSize | 82.5 MB | 82.5 MB | ✓ |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.047 | 2.038 | -0.01 |
| / avg req/sec | 1221.57 | 1226.85 | +5.28 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.078 | 1.102 | ⚠️ +0.02 |
| /error-in-render avg req/sec | 2319.51 | 2267.6 | ⚠️ -51.91 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| 677f882d2ed8..34b9.js gzip | 12.8 kB | 12.8 kB | ✓ |
| framework.HASH.js gzip | 39 kB | 39 kB | ✓ |
| main-3e8cb31..cddc.js gzip | 6.56 kB | 6.56 kB | ✓ |
| webpack-e067..f178.js gzip | 751 B | 751 B | ✓ |
| Overall change | 59 kB | 59 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| polyfills-4b..e242.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| _app-7231d4b..5856.js gzip | 1.28 kB | 1.28 kB | ✓ |
| _error-01375..90b6.js gzip | 3.69 kB | 3.69 kB | ✓ |
| hooks-d4591d..e7c2.js gzip | 887 B | 887 B | ✓ |
| index-17468f..5d83.js gzip | 227 B | 227 B | ✓ |
| link-db223d9..dbd7.js gzip | 1.61 kB | 1.61 kB | ✓ |
| routerDirect..fd29.js gzip | 303 B | 303 B | ✓ |
| withRouter-b..76ae.js gzip | 302 B | 302 B | ✓ |
| Overall change | 8.31 kB | 8.31 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| _buildManifest.js gzip | 322 B | 322 B | ✓ |
| Overall change | 322 B | 322 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| index.html gzip | 614 B | 614 B | ✓ |
| link.html gzip | 620 B | 620 B | ✓ |
| withRouter.html gzip | 608 B | 608 B | ✓ |
| Overall change | 1.84 kB | 1.84 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| buildDuration | 9.9s | 9.9s | -41ms |
| nodeModulesSize | 82.5 MB | 82.5 MB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| 677f882d2ed8..34b9.js gzip | 12.8 kB | 12.8 kB | ✓ |
| framework.HASH.js gzip | 39 kB | 39 kB | ✓ |
| main-3e8cb31..cddc.js gzip | 6.56 kB | 6.56 kB | ✓ |
| webpack-e067..f178.js gzip | 751 B | 751 B | ✓ |
| Overall change | 59 kB | 59 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| polyfills-4b..e242.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| _app-7231d4b..5856.js gzip | 1.28 kB | 1.28 kB | ✓ |
| _error-01375..90b6.js gzip | 3.69 kB | 3.69 kB | ✓ |
| hooks-d4591d..e7c2.js gzip | 887 B | 887 B | ✓ |
| index-17468f..5d83.js gzip | 227 B | 227 B | ✓ |
| link-db223d9..dbd7.js gzip | 1.61 kB | 1.61 kB | ✓ |
| routerDirect..fd29.js gzip | 303 B | 303 B | ✓ |
| withRouter-b..76ae.js gzip | 302 B | 302 B | ✓ |
| Overall change | 8.31 kB | 8.31 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| _buildManifest.js gzip | 322 B | 322 B | ✓ |
| Overall change | 322 B | 322 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| _error.js | 1 MB | 1 MB | ✓ |
| 404.html | 2.67 kB | 2.67 kB | ✓ |
| hooks.html | 1.92 kB | 1.92 kB | ✓ |
| index.js | 1 MB | 1 MB | ✓ |
| link.js | 1.06 MB | 1.06 MB | ✓ |
| routerDirect.js | 1.05 MB | 1.05 MB | ✓ |
| withRouter.js | 1.05 MB | 1.05 MB | ✓ |
| Overall change | 5.16 MB | 5.16 MB | ✓ |

</details>

</details>
Commit: 73763b7a3a4c20a01cbd344b59aff55f79538e4d

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| buildDuration | 8.2s | 8.5s | ⚠️ +306ms |
| nodeModulesSize | 82.5 MB | 82.5 MB | ✓ |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.158 | 2.19 | ⚠️ +0.03 |
| / avg req/sec | 1158.64 | 1141.68 | ⚠️ -16.96 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.116 | 1.131 | ⚠️ +0.01 |
| /error-in-render avg req/sec | 2239.73 | 2211.37 | ⚠️ -28.36 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| 677f882d2ed8..34b9.js gzip | 12.8 kB | 12.8 kB | ✓ |
| framework.HASH.js gzip | 39 kB | 39 kB | ✓ |
| main-3e8cb31..cddc.js gzip | 6.56 kB | 6.56 kB | ✓ |
| webpack-e067..f178.js gzip | 751 B | 751 B | ✓ |
| Overall change | 59 kB | 59 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| polyfills-4b..e242.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| _app-7231d4b..5856.js gzip | 1.28 kB | 1.28 kB | ✓ |
| _error-01375..90b6.js gzip | 3.69 kB | 3.69 kB | ✓ |
| hooks-d4591d..e7c2.js gzip | 887 B | 887 B | ✓ |
| index-17468f..5d83.js gzip | 227 B | 227 B | ✓ |
| link-db223d9..dbd7.js gzip | 1.61 kB | 1.61 kB | ✓ |
| routerDirect..fd29.js gzip | 303 B | 303 B | ✓ |
| withRouter-b..76ae.js gzip | 302 B | 302 B | ✓ |
| Overall change | 8.31 kB | 8.31 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| _buildManifest.js gzip | 322 B | 322 B | ✓ |
| Overall change | 322 B | 322 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| index.html gzip | 614 B | 614 B | ✓ |
| link.html gzip | 620 B | 620 B | ✓ |
| withRouter.html gzip | 608 B | 608 B | ✓ |
| Overall change | 1.84 kB | 1.84 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| buildDuration | 10s | 9.8s | -229ms |
| nodeModulesSize | 82.5 MB | 82.5 MB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| 677f882d2ed8..34b9.js gzip | 12.8 kB | 12.8 kB | ✓ |
| framework.HASH.js gzip | 39 kB | 39 kB | ✓ |
| main-3e8cb31..cddc.js gzip | 6.56 kB | 6.56 kB | ✓ |
| webpack-e067..f178.js gzip | 751 B | 751 B | ✓ |
| Overall change | 59 kB | 59 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| polyfills-4b..e242.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| _app-7231d4b..5856.js gzip | 1.28 kB | 1.28 kB | ✓ |
| _error-01375..90b6.js gzip | 3.69 kB | 3.69 kB | ✓ |
| hooks-d4591d..e7c2.js gzip | 887 B | 887 B | ✓ |
| index-17468f..5d83.js gzip | 227 B | 227 B | ✓ |
| link-db223d9..dbd7.js gzip | 1.61 kB | 1.61 kB | ✓ |
| routerDirect..fd29.js gzip | 303 B | 303 B | ✓ |
| withRouter-b..76ae.js gzip | 302 B | 302 B | ✓ |
| Overall change | 8.31 kB | 8.31 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| _buildManifest.js gzip | 322 B | 322 B | ✓ |
| Overall change | 322 B | 322 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong></summary>

|  | vercel/next.js canary  | xeoneux/next.js canary | Change |
| - | - | - | - |
| _error.js | 1 MB | 1 MB | ✓ |
| 404.html | 2.67 kB | 2.67 kB | ✓ |
| hooks.html | 1.92 kB | 1.92 kB | ✓ |
| index.js | 1 MB | 1 MB | ✓ |
| link.js | 1.06 MB | 1.06 MB | ✓ |
| routerDirect.js | 1.05 MB | 1.05 MB | ✓ |
| withRouter.js | 1.05 MB | 1.05 MB | ✓ |
| Overall change | 5.16 MB | 5.16 MB | ✓ |

</details>

</details>
Commit: b0a780fe02fe901209edd83e552168fdf5d19805

File: examples/with-pullstate/pages/_app.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+import { PullstateProvider } from 'pullstate'
+
+import { useHydrate } from '../stores'
+
+export default function App({ Component, pageProps }) {
+  const instance = useHydrate(pageProps.snapshot)
```
Comment (lfades) [inline examples/with-pullstate/pages/_app.js:6]: Let's give this another name, like `useStore`.

File: examples/with-pullstate/components/dark-mode-demo.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+      <button
+        onClick={() =>
+          UIStore.update((s) => {
+            s.isDarkMode = !s.isDarkMode
+          })
+        }
+      >
+        Toggle Dark Mode
+      </button>
+    </div>
```
Comment (lfades) [inline examples/with-pullstate/components/dark-mode-demo.js:27]: This component should have links to the other pages, that's important to see how all pages work together while the app is running.

**Review by lfades — CHANGES_REQUESTED** [review]
Added 2 more suggestions 🙏

**Comment by timneutkens** [comment]
Going to close this as the PR is stale. Happy to merge when the changes are made.

**Comment by Raamzeez** [comment]
Any progress made on this? Really would like to use Pullstate with Next.JS but there are no good examples.

**Comment by lfades** [comment]
@Raamzeez The PR was closed and examples are managed by the community, feel free to contribute

**Pull request closed without merging**